### PR TITLE
TS-1441 add last hit id support for tenures

### PR DESCRIPTION
**Pull request by LBHTKarki**
## Link to JIRA ticket

[TS-1441](https://hackney.atlassian.net/browse/TS-1441)

### *What is the problem we're trying to solve*

Bookings worktray used in Temporary Accommodation needs a last hit id based paged list of tenures sorted by tenure start date. Current tenure search end point does not support that functionality.

### *What changes have we introduced*

This update adds new `/tenures/all` end point that supports last hit id based paging while tenure start date based sorting is used at the same time. It supports the same search functionality as the current tenure search endpoint by using the same tenure query builder. Everything else has been kept separate for easier development further down the file. If the search functionality needs to be changed later new query builer can be implemented, so that these end points can be developed separately.

This update only covers the TA worktray sorting and paging (load more) use case and any other kind of paging and sorting by different values are not supported. They can be developed later as and when required.  

Basic usage for the new paging feature is:
`/all?searchText=%22%22&sortBy=tenureStartDate&isDesc=true&pageSize=7`

Updates include:

1. Add new `/all` end point to the `GetTenureListController`. The current implementation of `LambdaLogger` in the API does not really support testing, so that hasn't been covered. That's something we can look at improving API wide
2. Add e2e tests to cover the functionality of the new end point
3. Add test coverage for the `ElasticSearchWrapper` which currently does not have any tests. Approach taken with Elastic Search client is that the e2e tests are used to determine whether the client returns correct results rather than trying to mock the client and then verify that it's been called correctly. Those various extension methods would make mocking the client cumbersome and it's generally adviced against to try to mock it that way anyway
4. Addthe new `GetTenureListSetsUseCase` service to the Startup. Other changes in that file are purely formatting and sorting usings
5. Update `GetAllTenureListRequest` to include `LastHitTenureStartDate` property. When sorting is used in ES together with last hit id approach, both the id and the sorting value must be passed to the ES client. In this particular use case the LastHitTenureStartDate must be passed as well as the id of the last hit. Important thing to note about the date is that it must be provided in milliseconds since Unix Epoch. ES client returns the value in that format for the next set, so it's easy to pass it to the client in the same format
6. Add new `GetAllTenureListResponse` object that inherits from `GetTenureListResponse`
7. Add new `APIAllTenureResponse<T>` that inherits from `APIAllResponse<T>`. This way we don't have to pollute the `APIAllResponse` response with the `LastHitTenureStartDate` when it's not required. Such as in assets all response 
8. Add new `GetListOfTenuresSets` search gateway method.
9. Sort descriptor for `QueryableTenure` has been updated to include both tenure start date and last hit id. These both are required for ES to be able to work out what the last result set was and what should be returned next
10. Add `GetAllTenureListRequestValidator` boundary object validator to cover the new `LastHitTenureStartDate` property 

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Update the TA worktray to use the new end point.

[TS-1441]: https://hackney.atlassian.net/browse/TS-1441?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+        public void SetLastHitId(string lastHitId)
+        {
+            _lastHitId = lastHitId;
+        }
+
+        public string LastHitId()
+        {
+            return _lastHitId;
+        }
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs]: We want the search gateway to add the `LastHitId` and `LastHitTenureStartDate` to the `GetAllTenureListResponse` to be used later in the API response. However we don't want those properties to appear in the response inside `results` object, but at the top level.

The original get/set implementation followed the same pattern used elsewhere in response objects, but I've now updated it in the [#ea51f55](ea51f55a5f9302148f38d582c86e8f30b716da29) to ensure the properties only appear where we want while the implementation looks more like standard C# property handling. By making the properties internal in `GetAllTenureListResponse` they won't be included in the API response amongst the result automatically as it would be the case if they were public.

File: HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+namespace HousingSearchApi.V1.Boundary.Responses
+{
+    public class GetAllTenureListResponse : GetTenureListResponse
+    {
+        private string _lastHitId;
+        private string _lastHitTenureStartDate;
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs]: Please see above

File: HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Duslerke) [reply HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs]: Oh, I see.
In which case this can simplified a lot by simply adding the `[JsonIgnore]` [attributes](https://www.newtonsoft.com/json/help/html/PropertyJsonIgnore.htm) above the fields.

[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Responses/GetAllTenureListResponse.cs]: I was debating whether to go down that route, but because that approach wasn't used elsewhere in the API I initially decided againts it. However I do agree that using `[JsonIgnore]` makes things easier to manage and also makes it more obvious that we are doing that for "presentational" reasons. This approach is used in other APIs as well, so no reason to avoid it here really. I'll leave the other objects using private properties as is to keep the scope of these changes limited to this end point, but that's something we can look into changing if we have to update those models at some point. I have made the properties public in [#51b7b57](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/51b7b571f736d43ae01001e848f2a64fdfc923f5). Also tweaked the related test to be more specific. I did realise too that setting up the private properties like that is not necessary and sticking with the auto-implemented properties i.e. `internal string LastHitId { get; set; }` would have worked just the same. Anyway I think this option is better and makes it more obvious as to why we are handling those properties that way.

File: HousingSearchApi/V1/Controllers/GetTenureListController.cs [hunk 3]
```diff
@@ -45,5 +46,35 @@ public async Task<IActionResult> GetTenureList([FromQuery] GetTenureListRequest
...
+            {
+                var searchResults = await _getTenureListSetsUseCase.ExecuteAsync(request).ConfigureAwait(false);
+
+                var response = new APIAllTenureResponse<GetAllTenureListResponse>()
+                {
+                    Total = searchResults.Total(),
+                    LastHitId = searchResults.LastHitId(),
+                    LastHitTenureStartDate = searchResults.LastHitTenureStartDate(),
+                    Results = searchResults
+
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Controllers/GetTenureListController.cs]: removed in [#b5bb8c7](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/b5bb8c70fd5da7ed55a71fce6be7b24905e71367)

File: HousingSearchApi/V1/Gateways/SearchGateway.cs [hunk 4]
```diff
@@ -209,5 +209,39 @@ public async Task<List<Asset>> GetChildAssets(GetAssetRelationshipsRequest query
...
+                tenureListResponse
+                    .Tenures
+                    .AddRange(searchResponse.Documents.Select(queryableTenure => queryableTenure.Create()));
+
+                tenureListResponse.SetTotal(searchResponse.Total);
+
+                if (searchResponse.Documents.Count > 0)
+                {
+                    var lastHit = searchResponse.Hits.Last();
+                    var lastHitTenureStartDate = lastHit.Sorts?.Count > 0 ? lastHit.Sorts.First() : null;
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Gateways/SearchGateway.cs]: Updated in [#[b5bb8c7](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/b5bb8c70fd5da7ed55a71fce6be7b24905e71367)]. `Sorts` will typically be an empty array in ES response when sorting is not applied, so I've also updated the test setup to ensure it sets the `lastHitTenureStartDate` to null when `Sorts` is either null or empty array.

File: HousingSearchApi/V1/Boundary/Requests/GetAllTenureListRequest.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+using Microsoft.AspNetCore.Mvc;
+
+namespace HousingSearchApi.V1.Boundary.Requests
+{
+    public class GetAllTenureListRequest : GetTenureListRequest
+    {
+        [FromQuery(Name = "lastHitId")]
+        public string LastHitId { get; set; }
+
+        //in order to use sorting together with last hit id, this must be provided as well
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Requests/GetAllTenureListRequest.cs:10]: I've updated the validation in [#83d25a6](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/83d25a6a6d3ff630d84a212c80f20402ea4742f3), so that when `LastHitTenureStartDate` is provided then `LastHitId` must be provided as well. There's no dependency other way although the `lastHitId` in itself is not necessarily that useful without combining it with sorting in terms of getting predictable results. `LastHitId` can be used on its own however. I've also added validation for the `LastHitId` since it's the responsibility of this validator.

File: HousingSearchApi/V1/Boundary/Requests/GetAllTenureListRequest.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+namespace HousingSearchApi.V1.Boundary.Requests
+{
+    public class GetAllTenureListRequest : GetTenureListRequest
+    {
+        [FromQuery(Name = "lastHitId")]
+        public string LastHitId { get; set; }
+
+        //in order to use sorting together with last hit id, this must be provided as well 
+        [FromQuery(Name = "lastHitTenureStartDate")]
+        public string LastHitTenureStartDate { get; set; }
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Requests/GetAllTenureListRequest.cs:12]: Yes, `LastHitId` is the tenure id. The way this is designed to work is that when you use the `LastHitTenureStartDate` together with `LastHitId` style paging/load more feature the ES client provides the values for the next call. It always gives you the next two required values. These are available in the `Sort` array in the `Hit` object in the ES client response. So all the client has to do is to grab the values from the response and pass them back in the next call, so there's no need to calculate or fecth the start date at any point either on the API or the client. This offloads the responsibility fully to the ES client and since it's in charge of the values it wants that ensures things work as expected.

File: HousingSearchApi/V1/Boundary/Requests/Validation/GetAllTenureListRequestValidator.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+using FluentValidation;
+using Hackney.Core.Validation;
+
+namespace HousingSearchApi.V1.Boundary.Requests.Validation
+{
+    public class GetAllTenureListRequestValidator : AbstractValidator<GetAllTenureListRequest>
+    {
+        public GetAllTenureListRequestValidator()
+        {
+            When(x => !string.IsNullOrEmpty(x.LastHitTenureStartDate), () =>
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Boundary/Requests/Validation/GetAllTenureListRequestValidator.cs:11]: Please see the above comment re: dependencies between these two values and how that is now handled after latest update.

File: HousingSearchApi/V1/Gateways/SearchGateway.cs [hunk 8]
```diff
@@ -209,5 +209,39 @@ public async Task<List<Asset>> GetChildAssets(GetAssetRelationshipsRequest query
...
+            }
+            else
+            {
+                tenureListResponse
+                    .Tenures
+                    .AddRange(searchResponse.Documents.Select(queryableTenure => queryableTenure.Create()));
+
+                tenureListResponse.SetTotal(searchResponse.Total);
+
+                if (searchResponse.Documents.Count > 0)
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Gateways/SearchGateway.cs]: addressed in [#1618d3b](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/1618d3b60e5fb9d4d82b2ccbd9b9d57200792f25)

File: HousingSearchApi/V1/Infrastructure/ElasticElasticSearchWrapper.cs [hunk 9]
```diff
@@ -121,10 +122,60 @@ public class ElasticSearchWrapper : IElasticSearchWrapper
...
+
+            _logger.LogDebug($"ElasticSearch Search Tenures Sets begins {esNodes}");
+
+            if (request == null)
+            {
+                return new SearchResponse<QueryableTenure>();
+            }
+
+            //last hit paging is only supported when sorting by tenure start date
+            //any other type of paging is not supported. Page size can be set however when using start date sorting with last hit paging
```
[the remark this answers is not in the document]
Comment (LBHTKarki) [reply HousingSearchApi/V1/Infrastructure/ElasticElasticSearchWrapper.cs]: That's a good point about the note. I've updated the end point documentation in [#1618d3b](https://github.com/LBHackney-IT/housing-search-api/pull/226/commits/1618d3b60e5fb9d4d82b2ccbd9b9d57200792f25).

As for the functionality I was thinking about the same thing. However there's nothing TA specific here really since only enhancement is just the ability to be able to sort by tenureStartDate while using last hit id style paging. This could apply to other than TA use cases. For that reason I left it open for now. The way ES handles paging and sorting in that scenario means that it takes a bit of work to cover different use cases, say sorting by different or multiple properties, so this uptate only covers the bare minimum to achieve the functionality we need for the TA worktray. I did make the decision to not make  the gateway method generic for example, so it's tightly coupled to tenures, so there is some locking down happening already. That's not to say that it couldn't be written using generics, but it would required considerable amount of work and since we don't have any other use cases in mind right now I didn't think it was wort it.

**Review by Duslerke — APPROVED** [review]
Awesome! Look good

Thanks, for the in-depth answers on the comments/questions.
